modules/gateways/callback/coingate.php

```diff
@@ -1,69 +1,108 @@
-        break;
-    case 'canceled':
-        logTransaction($GATEWAY['name'], $response, 'Buyer canceled the invoice.');
-        break;
-    case 'expired':
-        logTransaction($GATEWAY['name'], $response, 'Buyer did not pay within the required time and the invoice expired.');
-        break;
-    case 'invalid':
-        logTransaction($GATEWAY['name'], $response, 'Payment rejected by the network or did not confirm.');
-        break;
-    case 'refunded':
-        logTransaction($GATEWAY['name'], $response, 'Payment was refunded to the buyer.');
-        break;
+$invoiceId = $_POST['order_id'];
+$transactionId = $_POST['id'];
+
+/**
+ * Validate Callback Invoice ID.
+ *
+ * Checks invoice ID is a valid invoice number. Note it will count an
+ * invoice in any status as valid.
+ *
+ * Performs a die upon encountering an invalid Invoice ID.
+ *
+ * Returns a normalised invoice ID.
+ *
+ * @param int $invoiceId Invoice ID
+ * @param string $gatewayName Gateway Name
+ */
+$invoiceId = checkCbInvoiceID($invoiceId, $gatewayParams['name']);
+
+/**
+ * Check Callback Transaction ID.
+ *
+ * Performs a check for any existing transactions with the same given
+ * transaction number.
+ *
+ * Performs a die upon encountering a duplicate.
+ *
+ * @param string $transactionId Unique Transaction ID
+ */
+checkCbTransID($transactionId);
+
+// ------------------------------------------------------------------
+// ------------------------------------------------------------------
+
+// Gateway Configuration Parameters
+$apiAuthToken = $gatewayParams['apiAuthToken'];
+$useSandboxEnv = $gatewayParams['useSandboxEnv'];
+
+require_once __DIR__ . '/../coingate/client.php';
+
+$client = new CoinGate\Client($apiAuthToken, $useSandboxEnv);
+
+$order = $client->order->get($transactionId);
+
+/**
+ * Log Transaction.
+ *
+ * Add an entry to the Gateway Log for debugging purposes.
+ *
+ * The debug data can be a string or an array. In the case of an
+ * array it will be
+ *
+ * @param string $gatewayName        Display label
+ * @param string|array $debugData    Data to log
+ * @param string $transactionStatus  Status
+ */
+
+$transactionStatusMessage = [
+    'paid' => 'Payment is confirmed by the network, and has been credited to the merchant. Purchased goods/services can be securely delivered to the buyer.',
+    'pending' => 'Buyer selected payment currency. Awaiting payment.',
+    'confirming' => 'Buyer transferred the payment for the invoice. Awaiting blockchain network confirmation.',
+    'canceled' => 'Buyer canceled the invoice.',
+    'expired' => 'Buyer did not pay within the required time and the invoice expired.',
+    'invalid' => 'Payment rejected by the network or did not confirm.',
+    'refunded' => 'Payment was refunded to the buyer.',
+];
+
+logTransaction($gatewayParams['name'], $_POST, $transactionStatusMessage[$order->status]);
```

Comment:
it is important not to overwite status, if the payment method was changed from coingate

Reply:
Could you get more in details about this?

What I did is just organized response status message translations in a cleaner way, e.g. see screenshot on how it was on the old implementation:
![image](https://user-images.githubusercontent.com/1407067/171373927-81e399e2-bee3-44b3-917a-ae8fd751fe26.png)

Reply:
we had the issue where user selects coingate as payment method,  returns back to checkout selects another payment method, pays the order.  but out callback expires the order after 20 minutes even even though another payment method was used? so some kind of check if still needed here 
`if (order.payment_method != 'coingate') { return }`

Reply:
It should not be an issue because "expire" does not set any order status on WHMCS (just logs under Billing > Gateway Log). So only when "paid" status is received from callback, WHMCS invoice is modified and market as "Complete".

Also, each callback for different "payment methods" are isolated and ``checkCbInvoiceID($order->order_id, $gatewayParams['name'])`` ensure and does exactly what you have asked.

Finally, I have adjusted flow to not trust ``invoiceId`` passed through $_POST for extra security but use one coming directly from SDK (fetched order object) and then validate it with ``checkCbInvoiceID``. This should be an additional flow upgrade from the old plugin.

Please check and let me know. :-)